src/cryptsetup/cryptsetup-tokens/cryptsetup-token-systemd-tpm2.c

```diff
@@ -156,45 +169,66 @@ _public_ void cryptsetup_token_dump(
                 struct crypt_device *cd /* is always LUKS2 context */,
                 const char *json /* validated 'systemd-tpm2' token if cryptsetup_token_validate is defined */) {
 
-        int r;
-        TPM2Flags flags = 0;
-        uint32_t pcr_mask;
+        _cleanup_free_ char *hash_pcrs_str = NULL, *pubkey_pcrs_str = NULL, *blob_str = NULL, *policy_hash_str = NULL, *pubkey_str = NULL;
+        _cleanup_free_ void *blob = NULL, *pubkey = NULL, *policy_hash = NULL;
+        _cleanup_(json_variant_unrefp) JsonVariant *v = NULL;
+        size_t blob_size, policy_hash_size, pubkey_size;
+        uint32_t hash_pcr_mask, pubkey_pcr_mask;
         uint16_t pcr_bank, primary_alg;
-        size_t decoded_blob_size;
-        _cleanup_free_ char *base64_blob = NULL, *hex_policy_hash = NULL,
-                            *pcrs_str = NULL, *blob_str = NULL, *policy_hash_str = NULL;
-        _cleanup_free_ void *decoded_blob = NULL;
+        TPM2Flags flags = 0;
+        int r;
 
         assert(json);
 
-        r = parse_luks2_tpm2_data(json, UINT32_MAX, &pcr_mask, &pcr_bank, &primary_alg, &base64_blob, &hex_policy_hash, &flags);
+        r = json_parse(json, 0, &v, NULL, NULL);
+        if (r < 0)
+                return (void) crypt_log_debug_errno(cd, r, "Failed to parse " TOKEN_NAME " JSON object: %m");
+
+        r = tpm2_parse_luks2_json(
```

Comment:
IIUC this generates a JSON that is shoved into the luks2 header so systemd can extract info later?

Reply:
This part *parses* the data from the LUKS2 header, and then uses it to unseal the LUKS key via the TPM2.

Reply:
Yeah I didn't mean to say generates (I was looking at A and commenting at B), but this is the data structure that gets generated and shoved into the LUKS header for systemd later? So when discussing verifying the session theirs an ESYS_TR I need to generate and storage would be via this JSON in the LUKS header?

Reply:
yupp